File: android/app/src/main/kotlin/au/radar/app/ui/SettingsScreen.kt
package au.radar.app.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.item
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Bolt
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.VolumeUp
import androidx.compose.material3.Icon
import androidx.compose.material3.Slider
import androidx.compose.material3.SliderDefaults
import androidx.compose.material3.Switch
import androidx.compose.material3.SwitchDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import au.radar.core.AlertSettings
import au.radar.core.KindSettings
import kotlin.math.roundToInt

/**
 * The whole settings surface, over the map like everything else.
 *
 * It is organised by what the driver cares about rather than by data model:
 * the two things people install this for sit at the top, cameras next, road
 * conditions after that, and the noisy stuff they will probably want quiet is
 * grouped under a heading that says so.
 */
/** One kind row, with the heading that precedes it when the group changes. */
private data class KindRowSpec(
    val kind: String,
    val title: String,
    val groupHeading: String?,
)

private val kindRows: List<KindRowSpec> = buildList {
    var previousGroup: String? = null
    for (entry in AlertSettings.editableKinds) {
        add(
            KindRowSpec(
                kind = entry.kind,
                title = entry.title,
                groupHeading = entry.group.takeIf { it != previousGroup },
            ),
        )
        previousGroup = entry.group
    }
}

@Composable
fun SettingsScreen(
    settings: AlertSettings,
    cameraCount: Int,
    connected: Boolean,
    onChange: (AlertSettings) -> Unit,
    onDismiss: () -> Unit,
) {
    fun patch(kind: String, block: (KindSettings) -> KindSettings) {
        val current = settings.forKind(kind)
        onChange(settings.copy(kinds = settings.kinds + (kind to block(current))))
    }

    Box(
        Modifier
            .fillMaxSize()
            .background(Color.Black.copy(alpha = 0.62f))
            .clickable(onClick = onDismiss),
    ) {
        Box(Modifier.padding(10.dp).clickable(enabled = false, onClick = {})) {
            GlassPanel(
                modifier = Modifier.fillMaxSize(),
                shape = RoundedCornerShape(28.dp),
                strong = true,
            ) {
                LazyColumn(
                    Modifier.fillMaxSize().padding(horizontal = 20.dp),
                    verticalArrangement = Arrangement.spacedBy(2.dp),
                ) {
                    item {
                        Row(
                            Modifier.fillMaxWidth().padding(top = 20.dp, bottom = 10.dp),
                            horizontalArrangement = Arrangement.SpaceBetween,
                            verticalAlignment = Alignment.CenterVertically,
                        ) {
                            Text(
                                "Alerts",
                                fontSize = 22.sp,
                                fontWeight = FontWeight.Bold,
                                color = Color.White,
                            )
                            GlassIconButton(
                                onClick = onDismiss,
                                contentDescription = "Close settings",
                                size = 40.dp,
                            ) { Icon(Icons.Filled.Close, contentDescription = null) }
                        }
                    }

                    item { SectionHeader("Overall") }

                    item {
                        ToggleRow(
                            title = "Spoken warnings",
                            subtitle = "Turn off for tones only",
                            checked = !settings.muted,
                            onCheckedChange = { onChange(settings.copy(muted = !it)) },
                        )
                    }
                    item {
                        ToggleRow(
                            title = "Flash the screen",
                            subtitle = "Pulses the edges — keeps working when muted",
                            checked = settings.flashEnabled,
                            onCheckedChange = { onChange(settings.copy(flashEnabled = it)) },
                        )
                    }
                    item {
                        SliderRow(
                            title = "Extra warning on my own road",
                            value = settings.sameRoadLeadMultiplier.toFloat(),
                            range = 1f..2.5f,
                            format = { "${(it * 100).roundToInt()}%" },
                            explanation = "Something on the road you are already on is " +
                                "warned about this much earlier than something merely in front.",
                            onChange = { onChange(settings.copy(sameRoadLeadMultiplier = it.toDouble())) },
                        )
                    }
                    item {
                        SliderRow(
                            title = "Stay quiet below",
                            value = settings.minSpeedKmh.toFloat(),
                            range = 0f..60f,
                            format = { if (it < 1) "Off" else "${it.roundToInt()} km/h" },
                            explanation = "No warnings at all under this speed.",
                            onChange = { onChange(settings.copy(minSpeedKmh = it.toDouble())) },
                        )
                    }

                    // Group headings are decided up front rather than tracked
                    // while composing: a lazy list builds items out of order and
                    // reuses them, so any running variable here would be wrong.
                    items(kindRows) { row ->
                        Column {
                            row.groupHeading?.let { SectionHeader(it) }
                            KindRow(
                                title = row.title,
                                kind = settings.forKind(row.kind),
                                onChange = { updated -> patch(row.kind) { updated } },
                            )
                        }
                    }

                    item { SectionHeader("This phone") }
                    item { InfoRow("Cameras stored", "$cameraCount") }
                    item { InfoRow("Live hazards", if (connected) "Connected" else "Offline") }

                    item {
                        Text(
                            "Camera locations come from state government open data. Live " +
                                "hazards come from the road authorities. This app never detects " +
                                "or interferes with any enforcement signal.",
                            color = Color.White.copy(0.45f),
                            fontSize = 11.sp,
                            modifier = Modifier.padding(vertical = 22.dp),
                        )
                    }
                }
            }
        }
    }
}

@Composable
private fun SectionHeader(text: String) {
    Text(
        text.uppercase(),
        color = Color.White.copy(0.42f),
        fontSize = 10.sp,
        fontWeight = FontWeight.SemiBold,
        letterSpacing = 1.4.sp,
        modifier = Modifier.padding(top = 22.dp, bottom = 8.dp),
    )
}

/**
 * One threat kind. Three controls, because those are the three questions a
 * driver actually has: do I want it, how loud, and how early.
 */
@Composable
private fun KindRow(
    title: String,
    kind: KindSettings,
    onChange: (KindSettings) -> Unit,
) {
    Column(Modifier.padding(vertical = 8.dp)) {
        Row(
            Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                title,
                color = if (kind.enabled) Color.White else Color.White.copy(0.4f),
                fontSize = 15.sp,
                fontWeight = FontWeight.Medium,
                modifier = Modifier.weight(1f),
            )
            Switch(
                checked = kind.enabled,
                onCheckedChange = { onChange(kind.copy(enabled = it)) },
                colors = switchColours(),
            )
        }

        if (kind.enabled) {
            Row(
                Modifier.padding(top = 6.dp),
                horizontalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                OptionChip(
                    label = "Voice",
                    icon = { Icon(Icons.Filled.VolumeUp, null, Modifier.size(14.dp)) },
                    selected = kind.voice,
                    onClick = { onChange(kind.copy(voice = !kind.voice)) },
                )
                OptionChip(
                    label = "Flash",
                    icon = { Icon(Icons.Filled.Bolt, null, Modifier.size(14.dp)) },
                    selected = kind.flash,
                    onClick = { onChange(kind.copy(flash = !kind.flash)) },
                )
            }

            SliderRow(
                title = "Warn me",
                value = kind.leadSeconds.toFloat(),
                range = 5f..90f,
                format = { "${it.roundToInt()} s ahead" },
                explanation = null,
                onChange = { onChange(kind.copy(leadSeconds = it.toDouble())) },
            )
            SliderRow(
                title = "Or anywhere within",
                value = kind.radiusM.toFloat(),
                range = 100f..1500f,
                format = { "${(it / 50).roundToInt() * 50} m" },
                explanation = null,
                onChange = { onChange(kind.copy(radiusM = it.toDouble())) },
            )
        }
    }
}

@Composable
private fun OptionChip(
    label: String,
    icon: @Composable () -> Unit,
    selected: Boolean,
    onClick: () -> Unit,
) {
    GlassPanel(
        modifier = Modifier.height(30.dp).clickable(onClick = onClick),
        shape = Glass.CapsuleShape,
        tint = if (selected) Color(0xFF2F80ED) else Color.Unspecified,
    ) {
        Row(
            Modifier.align(Alignment.Center).padding(horizontal = 12.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            icon()
            Spacer(Modifier.width(6.dp))
            Text(
                label,
                fontSize = 12.sp,
                color = if (selected) Color.White else Color.White.copy(0.55f),
            )
        }
    }
}

@Composable
private fun ToggleRow(
    title: String,
    subtitle: String,
    checked: Boolean,
    onCheckedChange: (Boolean) -> Unit,
) {
    Row(
        Modifier.fillMaxWidth().padding(vertical = 10.dp),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Column(Modifier.weight(1f)) {
            Text(title, color = Color.White, fontSize = 15.sp, fontWeight = FontWeight.Medium)
            Text(subtitle, color = Color.White.copy(0.5f), fontSize = 12.sp)
        }
        Switch(checked = checked, onCheckedChange = onCheckedChange, colors = switchColours())
    }
}

@Composable
private fun SliderRow(
    title: String,
    value: Float,
    range: ClosedFloatingPointRange<Float>,
    format: (Float) -> String,
    explanation: String?,
    onChange: (Float) -> Unit,
) {
    Column(Modifier.padding(top = 8.dp)) {
        Row(
            Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
        ) {
            Text(title, color = Color.White.copy(0.72f), fontSize = 13.sp)
            Text(
                format(value),
                color = Color.White,
                fontSize = 13.sp,
                fontWeight = FontWeight.SemiBold,
            )
        }
        Slider(
            value = value.coerceIn(range),
            onValueChange = onChange,
            valueRange = range,
            colors = SliderDefaults.colors(
                thumbColor = Color(0xFF4AA8FF),
                activeTrackColor = Color(0xFF2F80ED),
                inactiveTrackColor = Color.White.copy(0.18f),
            ),
        )
        if (explanation != null) {
            Text(explanation, color = Color.White.copy(0.42f), fontSize = 11.sp)
        }
    }
}

@Composable
private fun InfoRow(label: String, value: String) {
    Row(
        Modifier.fillMaxWidth().padding(vertical = 9.dp),
        horizontalArrangement = Arrangement.SpaceBetween,
    ) {
        Text(label, color = Color.White.copy(0.7f), fontSize = 14.sp)
        Text(value, color = Color.White, fontSize = 14.sp, fontWeight = FontWeight.Medium)
    }
}

@Composable
private fun switchColours() = SwitchDefaults.colors(
    checkedThumbColor = Color.White,
    checkedTrackColor = Color(0xFF2F80ED),
    uncheckedThumbColor = Color.White.copy(0.7f),
    uncheckedTrackColor = Color.White.copy(0.14f),
    uncheckedBorderColor = Color.White.copy(0.22f),
)

File: ios/App/Sources/SettingsScreen.swift
import RadarKit
import SwiftUI

/// The whole settings surface.
///
/// Organised by what the driver cares about rather than by data model: the two
/// things people install this for sit at the top, cameras next, road conditions
/// after that, and the noisy stuff they will probably want quiet is grouped
/// under a heading that says so.
struct SettingsScreen: View {
    @ObservedObject var model: DriveModel
    @Environment(\.dismiss) private var dismiss

    private var settings: AlertSettings { model.settings }

    /// The kind rows grouped for display, computed once rather than tracked
    /// while the list scrolls.
    private var groups: [(name: String, kinds: [(kind: String, title: String)])] {
        var ordered: [(String, [(String, String)])] = []
        for entry in AlertSettings.editableKinds {
            if let last = ordered.last, last.0 == entry.group {
                ordered[ordered.count - 1].1.append((entry.kind, entry.title))
            } else {
                ordered.append((entry.group, [(entry.kind, entry.title)]))
            }
        }
        return ordered.map { (name: $0.0, kinds: $0.1.map { (kind: $0.0, title: $0.1) }) }
    }

    var body: some View {
        NavigationStack {
            Form {
                Section("Overall") {
                    Toggle("Spoken warnings", isOn: Binding(
                        get: { !settings.muted },
                        set: { var next = settings; next.muted = !$0; model.update(settings: next) }
                    ))

                    Toggle("Flash the screen", isOn: Binding(
                        get: { settings.flashEnabled },
                        set: { var next = settings; next.flashEnabled = $0; model.update(settings: next) }
                    ))

                    LabelledSlider(
                        title: "Extra warning on my own road",
                        value: Binding(
                            get: { settings.sameRoadLeadMultiplier },
                            set: { var n = settings; n.sameRoadLeadMultiplier = $0; model.update(settings: n) }
                        ),
                        range: 1...2.5,
                        format: { "\(Int(($0 * 100).rounded()))%" }
                    )
                    Text(
                        "Something on the road you are already on is warned about this much "
                        + "earlier than something merely in front of you."
                    )
                    .font(.caption2)
                    .foregroundStyle(.secondary)

                    LabelledSlider(
                        title: "Stay quiet below",
                        value: Binding(
                            get: { settings.minSpeedKmh },
                            set: { var n = settings; n.minSpeedKmh = $0; model.update(settings: n) }
                        ),
                        range: 0...60,
                        format: { $0 < 1 ? "Off" : "\(Int($0.rounded())) km/h" }
                    )
                }

                ForEach(groups, id: \.name) { group in
                    Section(group.name) {
                        ForEach(group.kinds, id: \.kind) { entry in
                            KindEditor(
                                title: entry.title,
                                kind: settings.forKind(entry.kind),
                                onChange: { updated in
                                    var next = settings
                                    next.kinds[entry.kind] = updated
                                    model.update(settings: next)
                                }
                            )
                        }
                    }
                }

                Section {
                    Button {
                        model.plantTestCamera()
                        dismiss()
                    } label: {
                        Label("Plant a test camera 600 m ahead", systemImage: "camera.badge.ellipsis")
                    }
                    if !model.testThreats.isEmpty {
                        Button(role: .destructive) {
                            model.clearTestCameras()
                        } label: {
                            Label("Remove the test camera", systemImage: "trash")
                        }
                    }
                } header: {
                    Text("Testing")
                } footer: {
                    Text(
                        "Drops a pretend speed camera on the road ahead so you can hear "
                        + "and see a real warning without a backend. Drive towards it."
                    )
                }

                Section("This phone") {
                    LabeledContent("Cameras stored", value: "\(model.cameras.count)")
                    LabeledContent(
                        "Live hazards",
                        value: model.connected ? "Connected" : "Offline"
                    )
                }

                Section {
                    Text(
                        "Camera locations come from state government open data. Live hazards "
                        + "come from the road authorities. This app never detects or interferes "
                        + "with any enforcement signal."
                    )
                    .font(.caption2)
                    .foregroundStyle(.secondary)
                }
            }
            .navigationTitle("Alerts")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .confirmationAction) {
                    Button("Done") { dismiss() }
                }
            }
        }
    }
}

/// One threat kind. Three controls, because those are the three questions a
/// driver actually has: do I want it, how loud, and how early.
private struct KindEditor: View {
    let title: String
    let kind: KindSettings
    let onChange: (KindSettings) -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 10) {
            Toggle(title, isOn: Binding(
                get: { kind.enabled },
                set: { var next = kind; next.enabled = $0; onChange(next) }
            ))
            .font(.body.weight(.medium))

            if kind.enabled {
                HStack(spacing: 8) {
                    OptionChip(
                        title: "Voice",
                        systemImage: "speaker.wave.2.fill",
                        isOn: kind.voice
                    ) {
                        var next = kind; next.voice.toggle(); onChange(next)
                    }
                    OptionChip(
                        title: "Flash",
                        systemImage: "bolt.fill",
                        isOn: kind.flash
                    ) {
                        var next = kind; next.flash.toggle(); onChange(next)
                    }
                }

                LabelledSlider(
                    title: "Warn me",
                    value: Binding(
                        get: { kind.leadSeconds },
                        set: { var next = kind; next.leadSeconds = $0; onChange(next) }
                    ),
                    range: 5...90,
                    format: { "\(Int($0.rounded())) s ahead" }
                )

                LabelledSlider(
                    title: "Or anywhere within",
                    value: Binding(
                        get: { kind.radiusM },
                        set: { var next = kind; next.radiusM = $0; onChange(next) }
                    ),
                    range: 100...1500,
                    format: { "\(Int(($0 / 50).rounded()) * 50) m" }
                )
            }
        }
        .padding(.vertical, 4)
    }
}

private struct OptionChip: View {
    let title: String
    let systemImage: String
    let isOn: Bool
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Label(title, systemImage: systemImage)
                .font(.caption)
                .padding(.horizontal, 12)
                .padding(.vertical, 7)
                .contentShape(Capsule())
        }
        .buttonStyle(.plain)
        .foregroundStyle(isOn ? .white : .secondary)
        .radarGlassCapsule(tint: isOn ? RadarGlass.route : nil)
    }
}

private struct LabelledSlider: View {
    let title: String
    @Binding var value: Double
    let range: ClosedRange<Double>
    let format: (Double) -> String

    var body: some View {
        VStack(alignment: .leading, spacing: 2) {
            HStack {
                Text(title)
                    .font(.subheadline)
                    .foregroundStyle(.secondary)
                Spacer()
                Text(format(value))
                    .font(.subheadline.weight(.semibold))
                    .monospacedDigit()
            }
            Slider(value: $value, in: range)
        }
    }
}
